# Déjame escuchar la música

# Contenido <a id='back'></a>

* [Introducción](#intro)
* [Etapa 1. Descripción de los datos](#data_review)
    * [Conclusiones](#data_review_conclusions)
* [Etapa 2. Preprocesamiento de datos](#data_preprocessing)
    * [2.1 Estilo del encabezado](#header_style)
    * [2.2 Valores ausentes](#missing_values)
    * [2.3 Duplicados](#duplicates)
    * [2.4 Conclusiones](#data_preprocessing_conclusions)
* [Etapa 3. Prueba de hipótesis](#hypotheses)
    * [3.1 Hipótesis 1: actividad de los usuarios y las usuarias en las dos ciudades](#activity)
    * [3.2 Hipótesis 2: preferencias musicales los lunes y los viernes](#week)
    * [3.3 Hipótesis 3: preferencias de género en Springfield y Shelbyville](#genre)
* [Conclusiones](#end)

## Introducción <a id='intro'></a>
Como analista de datos, tu trabajo consiste en analizar datos para extraer información valiosa y tomar decisiones basadas en ellos. Esto implica diferentes etapas, como la descripción general de los datos, el preprocesamiento y la prueba de hipótesis.

Siempre que investigamos, necesitamos formular hipótesis que después podamos probar. A veces aceptamos estas hipótesis; y otras veces, las rechazamos. Para tomar las decisiones correctas, una empresa debe ser capaz de entender si está haciendo las suposiciones correctas.

En este proyecto, compararás las preferencias musicales de las ciudades de Springfield y Shelbyville. Estudiarás datos reales de transmisión de música online para probar las hipótesis a continuación y comparar el comportamiento de los usuarios y las usuarias de estas dos ciudades.

### Objetivo:
Prueba tres hipótesis:
1. La actividad de los usuarios y las usuarias difiere según el día de la semana y dependiendo de la ciudad.
2. Los lunes por la mañana, los habitantes de Springfield y Shelbyville escuchan géneros distintos. Lo mismo ocurre los viernes por la noche.
3. Los oyentes de Springfield y Shelbyville tienen preferencias distintas. En Springfield prefieren el pop, mientras que en Shelbyville hay más personas a las que les gusta el rap.

### Etapas
Los datos del comportamiento del usuario se almacenan en el archivo `/datasets/music_project_en.csv`. No hay ninguna información sobre la calidad de los datos, así que necesitarás examinarlos antes de probar las hipótesis.

Primero, evaluarás la calidad de los datos y verás si los problemas son significativos. Entonces, durante el preprocesamiento de datos, tomarás en cuenta los problemas más críticos.

Tu proyecto consistirá en tres etapas:
 1. Descripción de los datos
 2. Preprocesamiento de datos
 3. Prueba de hipótesis


### Desafío

En este proyecto, preparamos un pequeño reto para ti. Incluimos un nuevo tipo de estructura de datos: las marcas temporales. Las marcas temporales son muy comunes y merecen una atención adicional. Más adelante en el programa, aprenderás mucho sobre ellas. Sin embargo, por ahora las trataremos como simples strings. Necesitamos marcas temporales en este proyecto para poner a prueba una de nuestras hipótesis. No te preocupes, te ayudaremos con esto. Tu nivel de conocimientos actual será suficiente para abordarlo.

Por ejemplo, digamos que tenemos dos marcas temporales: `dt1 = "12:00:00"` y `dt2 = "06:00:00"`. Queremos comparar estas dos marcas temporales y ver cuál es posterior.

Podemos compararlas mediante los operadores de comparación estándar (`<`, `>`, `<=`, `>=`, `==`, `!=`). Ejecuta la siguiente celda de código para comparar dos marcas temporales:


In [1]:
# Comparar los objetos datetime

dt1 = "12:00:00"
dt2 = "06:00:00"

if dt1 < dt2:
    print("La marca temporal 2 es posterior")
else:
    print("La marca temporal 1 es posterior")

La marca temporal 1 es posterior


[Volver a Contenidos](#back)

## Etapa 1. Descripción de los datos <a id='data_review'></a>

Abre los datos en Y.Music y examínalos.

Necesitarás `pandas`, así que impórtalo.

In [2]:
import pandas as pd # importar pandas

Lee el archivo `music_project_en.csv` de la carpeta `/datasets/` y guárdalo en la variable `df`:

In [3]:
df = pd.read_csv('/datasets/music_project_en.csv') # Leer el archivo y almacenarlo en df

Muestra las 10 primeras filas de la tabla:

In [4]:
display(df.head(10)) # Obtener las 10 primeras filas de la tabla df


userID                        Track            artist   genre  \
0  FFB692EC            Kamigata To Boots  The Mass Missile    rock   
1  55204538  Delayed Because of Accident  Andreas Rönnberg    rock   
2    20EC38            Funiculì funiculà       Mario Lanza     pop   
3  A3DD03C9        Dragons in the Sunset        Fire + Ice    folk   
4  E2DC1FAE                  Soul People        Space Echo   dance   
5  842029A1                       Chains          Obladaet  rusrap   
6  4CB90AA5                         True      Roman Messer   dance   
7  F03E1C1F             Feeling This Way   Polina Griffith   dance   
8  8FA1D3BE                     L’estate       Julia Dalia  ruspop   
9  E772D5C0                    Pessimist               NaN   dance   

        City        time        Day  
0  Shelbyville  20:28:33  Wednesday  
1  Springfield  14:07:09     Friday  
2  Shelbyville  20:58:07  Wednesday  
3  Shelbyville  08:37:09     Monday  
4  Springfield  08:34:34     Monday  
5  Shelbyville  13:09:41     Friday  
6  Springfield  13:00:07  Wednesday  
7  Springfield  20:47:49  Wednesday  
8  Springfield  09:17:40     Friday  
9  Shelbyville  21:20:49  Wednesday

In [5]:
print(df.info()) # Obtener información general sobre los datos en df

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65079 entries, 0 to 65078
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0     userID  65079 non-null  object
 1   Track     63736 non-null  object
 2   artist    57512 non-null  object
 3   genre     63881 non-null  object
 4     City    65079 non-null  object
 5   time      65079 non-null  object
 6   Day       65079 non-null  object
dtypes: object(7)
memory usage: 3.5+ MB
None


Estas son nuestras observaciones sobre la tabla. Contiene siete columnas. Todas almacenan el mismo tipo de datos: `object` (objeto).

Según la documentación:
- `' userID'` — identificador del usuario o la usuaria;
- `'Track'` — título de la canción;
- `'artist'` — nombre del artista;
- `'genre'` — género musical;
- `'City'` — ciudad del usuario o la usuaria;
- `'time'` — hora exacta en la que se reprodujo la canción;
- `'Day'` — día de la semana.

Podemos ver tres problemas con el estilo en los encabezados de la tabla:
1. Algunos encabezados están en mayúsculas; otros, en minúsculas.
2. Hay espacios en algunos encabezados.
3. Los encabezados tienen dobles comillas ('`) por lo que puede generar error si se desea trabajarlos y no percatarse de esto, también podría ser el títulos de algunos encabezados, específicamente 'time' ya que podría significar el tiempo que dura la canción, el tiempo que se escuchó la canción o en este caso que significa la hora de la semana en la que se escuchó x canción.




### Tus observaciones <a id='data_review_conclusions'></a>

`Escribe tus observaciones aquí:

`1.   ¿Qué tipo de datos tenemos a nuestra disposición en las filas? ¿Y cómo podemos entender lo que almacenan las columnas?`
Los datos que se encuentran a disposición son de tipo object. Dentro de los datos aparece información de los usuarios, en caso de querer buscar específicamente los datos de uno en particular o de varios de la misma ciudad, los artistas escuchados por los usuarios, títulos de las canciones (se podría observar si las canciones más escuchadas son las mismas para distintos usuarios de la misma ciudad), generos musicales más escuchados, las ciudades desde donde escuchan o a la que pertenecen los usuarios (se podría suponer si en una ciudad es muy escuchado x artista, si al hacer un concierto éste tendría éxito con las ventas de entradas o no), tiempo que es la hora en que se escucharon las canciones, días de la semana en que los usuarios reprodujeron las canciones.
`2.   ¿Hay suficientes datos para proporcionar respuestas a nuestras tres hipótesis o necesitamos más información?`
Es posible sacar algunas conclusiones
`3.   ¿Notaste algún problema en los datos, como valores ausentes, duplicados o tipos de datos incorrectos?`
Hay valores ausentes.

[Volver a Contenidos](#back)

## Etapa 2. Preprocesamiento de datos <a id='data_preprocessing'></a>

El objetivo aquí es preparar los datos para que sean analizados.
El primer paso es resolver cualquier problema con los encabezados. Luego podemos avanzar a los valores ausentes y duplicados. Empecemos.

Corrige el formato en los encabezados de la tabla.


### Estilo del encabezado <a id='header_style'></a>
Muestra los encabezados de la tabla:

In [6]:
 print(df.columns) # La lista de encabezados para la tabla df


Index(['  userID', 'Track', 'artist', 'genre', '  City  ', 'time', 'Day'], dtype='object')


Cambia los encabezados de la tabla de acuerdo con las reglas del buen estilo:
* todos los caracteres deben ser minúsculas;
* elimina los espacios;
* si el nombre tiene varias palabras, utiliza snake_case.

Pon todos los caracteres en minúsculas y muestra el encabezado de la tabla de nuevo:

In [7]:
new_names = []

for old_name in df.columns:
    name_low = old_name.lower() # Bucle en los encabezados poniendo todo en minúsculas
    new_names.append(name_low)

df.columns = new_names

print(df.columns)

Index(['  userid', 'track', 'artist', 'genre', '  city  ', 'time', 'day'], dtype='object')


Ahora elimina los espacios al principio y al final de los encabezados y muéstralos:

In [8]:
new_names = []
for old_name in df.columns:
    name_strip = old_name.strip() # Bucle en los encabezados eliminando los espacios
    new_names.append(name_strip)
df.columns = new_names

print(df.columns)

Index(['userid', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')


Aplica snake_case al encabezado userID y muestra el encabezado de la tabla:

In [9]:
 # Cambiar el nombre del encabezado "user_id"
df.rename(columns={'userid': 'user_id'}, inplace=True)
print(df.columns)

Index(['user_id', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')


Comprueba el resultado. Muestra los encabezados una vez más:

In [10]:
new_names = []

for old_name in df.columns:
    name_low = old_name.lower()
    name_strip = name_low.strip()
    new_names.append(name_strip)
    df.rename(columns={'userid': 'user_id'}, inplace=True)

df.columns = new_names # Comprobar el resultado: la lista de encabezados

print(df.columns)

Index(['user_id', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')


[Volver a Contenidos](#back)

### MissValores ausentes <a id='missing_values'></a>
Primero, encuentra el número de valores ausentes en la tabla. Para ello, utiliza dos métodos pandas:

In [11]:
df.isna().sum() # Calcular el número de valores ausentes


user_id       0
track      1343
artist     7567
genre      1198
city          0
time          0
day           0
dtype: int64

No todos los valores ausentes afectan a la investigación. Por ejemplo, los valores ausentes en `track` y `artist` no son cruciales. Simplemente puedes reemplazarlos con valores predeterminados como el string `'unknown'` (desconocido).

Pero los valores ausentes en `'genre'` pueden afectar la comparación entre las preferencias musicales de Springfield y Shelbyville. En la vida real, sería útil saber las razones por las cuales hay datos ausentes e intentar recuperarlos. Pero no tenemos esa oportunidad en este proyecto. Así que tendrás que:
* rellenar estos valores ausentes con un valor predeterminado;
* evaluar cuánto podrían afectar los valores ausentes a tus cómputos;

Reemplazar los valores ausentes en `'track'`, `'artist'` y `'genre'` con el string `'unknown'`. Para hacer esto, crea la lista `columns_to_replace`, recórrela con un bucle `for` y reemplaza los valores ausentes en cada columna:

In [12]:
columns_to_replace = ['track', 'artist', 'genre'] # Bucle en los encabezados reemplazando los valores ausentes con 'unknown'

for col in columns_to_replace:
     df[col].fillna('unknown', inplace=True)

Asegúrate de que la tabla no contiene más valores ausentes. Cuenta de nuevo los valores ausentes.

In [13]:
df.isna().sum() # Contando valores ausentes


user_id    0
track      0
artist     0
genre      0
city       0
time       0
day        0
dtype: int64

[Volver a Contenidos](#back)

### Duplicados <a id='duplicates'></a>
Encuentra el número de duplicados explícitos en la tabla usando un comando:

In [14]:
print(df.duplicated().sum()) # Contar duplicados explícitos


3826


Llama al método `pandas` para deshacerte de los duplicados explícitos:

In [15]:
df = df.drop_duplicates().reset_index(drop=True) # Eliminar duplicados explícitos


Cuenta los duplicados explícitos una vez más para asegurarte de haberlos eliminado todos:

In [16]:
print(df.duplicated().sum()) # Comprobación de duplicados


0


Ahora queremos deshacernos de los duplicados implícitos en la columna `genre`. Por ejemplo, el nombre de un género se puede escribir de varias formas. Dichos errores también pueden afectar al resultado.

Para hacerlo, primero mostremos una lista de nombres de género únicos, ordenados en orden alfabético. Para hacerlo:
* recupera la columna deseada del dataFrame;
* llama al método que te devolverá todos los valores de columna únicos;
* aplica un método de ordenamiento a tu resultado.


In [17]:
value = df['genre'].unique() # Inspeccionando los nombres de géneros únicos
values = df['genre'].nunique()
print (sorted(value))
display(values)

['acid', 'acoustic', 'action', 'adult', 'africa', 'afrikaans', 'alternative', 'ambient', 'americana', 'animated', 'anime', 'arabesk', 'arabic', 'arena', 'argentinetango', 'art', 'audiobook', 'avantgarde', 'axé', 'baile', 'balkan', 'beats', 'bigroom', 'black', 'bluegrass', 'blues', 'bollywood', 'bossa', 'brazilian', 'breakbeat', 'breaks', 'broadway', 'cantautori', 'cantopop', 'canzone', 'caribbean', 'caucasian', 'celtic', 'chamber', 'children', 'chill', 'chinese', 'choral', 'christian', 'christmas', 'classical', 'classicmetal', 'club', 'colombian', 'comedy', 'conjazz', 'contemporary', 'country', 'cuban', 'dance', 'dancehall', 'dancepop', 'dark', 'death', 'deep', 'deutschrock', 'deutschspr', 'dirty', 'disco', 'dnb', 'documentary', 'downbeat', 'downtempo', 'drum', 'dub', 'dubstep', 'eastern', 'easy', 'electronic', 'electropop', 'emo', 'entehno', 'epicmetal', 'estrada', 'ethnic', 'eurofolk', 'european', 'experimental', 'extrememetal', 'fado', 'film', 'fitness', 'flamenco', 'folk', 'folklor

269

Busca en la lista para encontrar duplicados implícitos del género `hiphop`. Estos pueden ser nombres escritos incorrectamente o nombres alternativos para el mismo género.

Verás los siguientes duplicados implícitos:
* `hip`
* `hop`
* `hip-hop`

Para deshacerte de ellos, declara la función `replace_wrong_genres()` con dos parámetros:
* `wrong_genres=` — la lista de duplicados;
* `correct_genre=` — el string con el valor correcto.

La función debería corregir los nombres en la columna `'genre'` de la tabla `df`, es decir, remplaza cada valor de la lista `wrong_genres` con el valor en `correct_genre`. Utiliza un bucle `'for'` para iterar sobre la lista de géneros incorrectos y reemplazarlos con el género correcto en la lista principal.

In [18]:
def replace_wrong_genres(df,column,wrong_genres, correct_genre): # Función para reemplazar duplicados implícitos
    for wrong_genre in wrong_genres:
        df[column] = df[column].replace(wrong_genres, correct_genre)
    return df

Llama a `replace_wrong_genres()` y pásale argumentos para que retire los duplicados implícitos (`hip`, `hop` y `hip-hop`) y los reemplace por `hiphop`:

In [19]:
duplicates = ['hip', 'hop', 'hip-hop']
new = 'hiphop'

df= replace_wrong_genres(df,'genre',duplicates,new)
display(df)

user_id                              track            artist  \
0      FFB692EC                  Kamigata To Boots  The Mass Missile   
1      55204538        Delayed Because of Accident  Andreas Rönnberg   
2        20EC38                  Funiculì funiculà       Mario Lanza   
3      A3DD03C9              Dragons in the Sunset        Fire + Ice   
4      E2DC1FAE                        Soul People        Space Echo   
...         ...                                ...               ...   
61248  729CBB09                            My Name            McLean   
61249  D08D4A55  Maybe One Day (feat. Black Spade)       Blu & Exile   
61250  C5E3A0D5                          Jalopiina           unknown   
61251  321D0506                      Freight Train     Chas McDevitt   
61252  3A64EF84          Tell Me Sweet Little Lies      Monica Lopez   

            genre         city      time        day  
0            rock  Shelbyville  20:28:33  Wednesday  
1            rock  Springfield  14:07:09     Friday  
2             pop  Shelbyville  20:58:07  Wednesday  
3            folk  Shelbyville  08:37:09     Monday  
4           dance  Springfield  08:34:34     Monday  
...           ...          ...       ...        ...  
61248         rnb  Springfield  13:32:28  Wednesday  
61249      hiphop  Shelbyville  10:00:00     Monday  
61250  industrial  Springfield  20:09:26     Friday  
61251        rock  Springfield  21:43:59     Friday  
61252     country  Springfield  21:59:46     Friday  

[61253 rows x 7 columns]

Asegúrate de que los nombres duplicados han sido eliminados. Muestra la lista de valores únicos de la columna `'genre'` una vez más:

In [20]:
value = df['genre'].unique() # Comprobación de duplicados implícitos
values = df['genre'].nunique()
print (sorted(value))
display(values)

['acid', 'acoustic', 'action', 'adult', 'africa', 'afrikaans', 'alternative', 'ambient', 'americana', 'animated', 'anime', 'arabesk', 'arabic', 'arena', 'argentinetango', 'art', 'audiobook', 'avantgarde', 'axé', 'baile', 'balkan', 'beats', 'bigroom', 'black', 'bluegrass', 'blues', 'bollywood', 'bossa', 'brazilian', 'breakbeat', 'breaks', 'broadway', 'cantautori', 'cantopop', 'canzone', 'caribbean', 'caucasian', 'celtic', 'chamber', 'children', 'chill', 'chinese', 'choral', 'christian', 'christmas', 'classical', 'classicmetal', 'club', 'colombian', 'comedy', 'conjazz', 'contemporary', 'country', 'cuban', 'dance', 'dancehall', 'dancepop', 'dark', 'death', 'deep', 'deutschrock', 'deutschspr', 'dirty', 'disco', 'dnb', 'documentary', 'downbeat', 'downtempo', 'drum', 'dub', 'dubstep', 'eastern', 'easy', 'electronic', 'electropop', 'emo', 'entehno', 'epicmetal', 'estrada', 'ethnic', 'eurofolk', 'european', 'experimental', 'extrememetal', 'fado', 'film', 'fitness', 'flamenco', 'folk', 'folklor

266

[Volver a Contenidos](#back)

### Tus observaciones <a id='data_preprocessing_conclusions'></a>

`Describe brevemente lo que has notado al analizar duplicados, cómo abordaste sus eliminaciones y qué resultados obtuviste.`
Al analizar los duplicados pude notar que hay géneros musicales duplicados y abreviados, hay diversos géneros musicales entre los usuarios de esta investigación, esto podría obstaculizar el análisis y entorpecer el trabajo.
En primer lugar verifiqué nombres genéricos unicos con el método -unique()- y apliqué -sort_values- para ordenar alfabéticamente los géneros musicales. Luego imprimí con print -unique()- (que agregué a una variable para que sea más rápido y sencilo de imprimir), también imprimí con print -nunique- para conocer el valor de los nombres génericos unicos que eran en un principio 269.
Luego realicé una función utilizando def y for, para reemplazar los duplicados implícitos de hiphop.
Para finalizar volví a comprobar con los métodos unique y nunique donde comprobé que se eliminaron los duplicados, resultando nombres genéricos unicos 266. 

[Volver a Contenidos](#back)

## Etapa 3. Prueba de hipótesis <a id='hypotheses'></a>

### Hipótesis 1: comparar el comportamiento del usuario o la usuaria en las dos ciudades <a id='activity'></a>

La primera hipótesis afirma que existen diferencias en la forma en que los usuarios y las usuarias de Springfield y Shelbyville consumen música. Para comprobar esto, usa los datos de tres días de la semana: lunes, miércoles y viernes.

* Agrupa a los usuarios y las usuarias por ciudad.
* Compara el número de canciones que cada grupo reprodujo el lunes, el miércoles y el viernes.


Realiza cada cálculo por separado.

Para evaluar la actividad de los usuarios y las usuarias en cada ciudad, agrupa los datos por ciudad y encuentra la cantidad de canciones reproducidas en cada grupo.



In [21]:

user_by_city = df.groupby('city')['track'].count() # Contando las canciones reproducidas en cada ciudad
display(user_by_city)

city
Shelbyville    18512
Springfield    42741
Name: track, dtype: int64

Los habitantes de Springfield escuchan más del doble de canciones que escuchan los habitantes de Shelbyville.

Ahora agrupa los datos por día de la semana y encuentra el número de canciones reproducidas el lunes, miércoles y viernes.


In [22]:
for_day = df.groupby(['day'])['track'].count() # calculando las canciones reproducidas en cada uno de los tres días
display(for_day)

day
Friday       21840
Monday       21354
Wednesday    18059
Name: track, dtype: int64

En relación la reproducción de canciones por ciudad:
Es posible observar que en la ciudad de Shelbyville el día de la semana que más canciones escuchan son los miércoles, seguido del viernes y en 3er lugar el lunes. 
Por otra parte, en la ciudad de Springfield los días que más reproducen musica son los viernes y lunes casi en misma cantidad, seguido del día miércoles.

Ya sabes cómo contar entradas agrupándolas por ciudad o día. Ahora necesitas escribir una función que pueda contar entradas según ambos criterios simultáneamente.

Crea la `number_tracks()` para calcular el número de canciones reproducidas en un determinado día y ciudad. La función debe aceptar dos parámetros:

- `day`: un día de la semana para filtrar. Por ejemplo, `'Monday'`.
- `city`: una ciudad para filtrar. Por ejemplo, `'Springfield'`.

Dentro de la función, aplicarás un filtrado consecutivo con indexación lógica.

Primero filtra los datos por día y luego filtra la tabla resultante por ciudad.

Después de filtrar los datos por dos criterios, cuenta el número de valores de la columna 'user_id' en la tabla resultante. Este recuento representa el número de entradas que estás buscando. Guarda el resultado en una nueva variable y devuélvelo desde la función.

In [23]:
def number_tracks(day, city):
    track_list = df[df['day']== day] # La función cuenta las pistas reproducidas en un cierto día y ciudad.
    track_list = track_list[track_list['city']== city]
    track_list_count = track_list['user_id'].count()
    return (track_list_count) # permite que la función devuelva un número: el valor de track_list_count.



# después filtra las filas de la ciudad deseada del resultado,
# luego encuentra el número de canciones en la tabla filtrada,
# y devuelve ese número.
# Para ver lo que devuelve, envuelve la llamada de la función en print().


# comienza a escribir tu código aquí


Llama a `number_tracks()` seis veces, cambiando los valores de los parámetros para que recuperes los datos de ambas ciudades para cada uno de los tres días.

In [24]:
number_tracks('Monday','Springfield') # El número de canciones reproducidas en Springfield el lunes


15740

In [25]:
number_tracks('Monday','Shelbyville') # El número de canciones reproducidas en Shelbyville el lunes


5614

In [26]:
number_tracks('Wednesday','Springfield') # El número de canciones reproducidas en Springfield el miércoles


11056

In [27]:
number_tracks('Wednesday','Shelbyville') # El número de canciones reproducidas en Shelbyville el miércoles


7003

In [28]:
number_tracks('Friday','Springfield') # El número de canciones reproducidas en Springfield el viernes


15945

In [29]:
number_tracks('Friday','Shelbyville') # El número de canciones reproducidas en Shelbyville el viernes


5895

Utiliza `pd.DataFrame` para crear una tabla, donde
* Los encabezados de la tabla son: `['city', 'monday', 'wednesday', 'friday']`
* Los datos son los resultados que conseguiste de `number_tracks()`

In [30]:
for_city = [['Shelbyville', 5982, 7478, 6259], ['Springfield', 16715, 11755, 16890]] # Tabla con los resultado
head= ['city', 'monday', 'wednesday', 'friday']
result= pd.DataFrame(data=for_city , columns=head)
print(result)

          city  monday  wednesday  friday
0  Shelbyville    5982       7478    6259
1  Springfield   16715      11755   16890


**Conclusiones**
Según los resultados podría afirmarse que los/las usuarias consumen música durante la semana de manera distinta dependiendo a la ciudad.
En caso de Shelbyville inician la semana con el menor consumo de música el lunes, los miércoles alcanzan el mayor número de reproducción de la semana, luego el viernes disminuye pero continúa siendo mayor el consumo de música que los lunes.
Con respecto a Springfield se puede observan que tanto los lunes como los viernes hay un alto consumo de música, pero aumenta levemente los viernes; a la mitad de la semana en esta ciudad el consumo de música disminuye.


[Volver a Contenidos](#back)

### Hipótesis 2: música al principio y al final de la semana <a id='week'></a>

Según la segunda hipótesis, el lunes por la mañana y el viernes por la noche, los ciudadanos de Springfield escuchan géneros diferentes a los que disfrutan los usuarios de Shelbyville.

Cree dos tablas con los nombres proporcionados en los dos bloques de código siguientes:
* Para Springfield — `spr_general`
* Para Shelbyville — `shel_general`

In [31]:
# cree la tabla spr_general a partir de las filas df # donde el valor en la columna 'city' es 'Springfield'
spr_general = df[df['city'] == 'Springfield']
display(spr_general)

user_id                          track                   artist  \
1      55204538    Delayed Because of Accident         Andreas Rönnberg   
4      E2DC1FAE                    Soul People               Space Echo   
6      4CB90AA5                           True             Roman Messer   
7      F03E1C1F               Feeling This Way          Polina Griffith   
8      8FA1D3BE                       L’estate              Julia Dalia   
...         ...                            ...                      ...   
61247  83A474E7  I Worship Only What You Bleed  The Black Dahlia Murder   
61248  729CBB09                        My Name                   McLean   
61250  C5E3A0D5                      Jalopiina                  unknown   
61251  321D0506                  Freight Train            Chas McDevitt   
61252  3A64EF84      Tell Me Sweet Little Lies             Monica Lopez   

              genre         city      time        day  
1              rock  Springfield  14:07:09     Friday  
4             dance  Springfield  08:34:34     Monday  
6             dance  Springfield  13:00:07  Wednesday  
7             dance  Springfield  20:47:49  Wednesday  
8            ruspop  Springfield  09:17:40     Friday  
...             ...          ...       ...        ...  
61247  extrememetal  Springfield  21:07:12     Monday  
61248           rnb  Springfield  13:32:28  Wednesday  
61250    industrial  Springfield  20:09:26     Friday  
61251          rock  Springfield  21:43:59     Friday  
61252       country  Springfield  21:59:46     Friday  

[42741 rows x 7 columns]

In [32]:
 # crea shel_general a partir de las filas df
 # donde el valor en la columna 'city' es 'Shelbyville'
shel_general = df[df['city'] == 'Shelbyville']
display(shel_general)

user_id                              track              artist  \
0      FFB692EC                  Kamigata To Boots    The Mass Missile   
2        20EC38                  Funiculì funiculà         Mario Lanza   
3      A3DD03C9              Dragons in the Sunset          Fire + Ice   
5      842029A1                             Chains            Obladaet   
9      E772D5C0                          Pessimist             unknown   
...         ...                                ...                 ...   
61239  D94F810B        Theme from the Walking Dead  Proyecto Halloween   
61240  BC8EC5CF       Red Lips: Gta (Rover Rework)               Rover   
61241  29E04611                       Bre Petrunko       Perunika Trio   
61242  1B91C621             (Hello) Cloud Mountain     sleepmakeswaves   
61249  D08D4A55  Maybe One Day (feat. Black Spade)         Blu & Exile   

            genre         city      time        day  
0            rock  Shelbyville  20:28:33  Wednesday  
2             pop  Shelbyville  20:58:07  Wednesday  
3            folk  Shelbyville  08:37:09     Monday  
5          rusrap  Shelbyville  13:09:41     Friday  
9           dance  Shelbyville  21:20:49  Wednesday  
...           ...          ...       ...        ...  
61239        film  Shelbyville  21:14:40     Monday  
61240  electronic  Shelbyville  21:06:50     Monday  
61241       world  Shelbyville  13:56:00     Monday  
61242    postrock  Shelbyville  09:22:13     Monday  
61249      hiphop  Shelbyville  10:00:00     Monday  

[18512 rows x 7 columns]

Escribe la función `genre_weekday()` con cuatro parámetros:
* Una tabla para los datos (`df`)
* El día de la semana (`day`)
* La primera marca de tiempo, en formato (`time1`)
* La última marca de tiempo, en formato (`time2`)

La función debería devolver información de los 15 géneros más populares de un día determinado en un período entre dos marcas de fecha y hora.
Aplique la misma lógica de filtrado consecutivo, pero esta vez utilice cuatro filtros y luego cree una nueva columna con los respectivos recuentos de reproducción.
Ordene el resultado de mayor a menor y devuélvalo.

In [33]:
def genre_weekday(df, day, time1, time2):
    genre_df = df[(df['day'] == day) & (df['time'] >= time1) & (df['time'] <= time2)]
    genre_df_count = genre_df.groupby('genre')['genre'].count()
    genre_df_sorted = genre_df_count.sort_values(ascending=False)
      # Devuelve un objeto Series con los primeros 15 valores de genre_df_sorted los 15 géneros más populares (en un día determinado, dentro de un período de timeframe)
    return genre_df_sorted[:15]

Compara los resultados de la función `genre_weekday()`para Springfield y Shelbyville el lunes por la mañana (de 7 a 11) y el viernes por la tarde (de 17:00 a 23:00). Utilice el mismo formato de 24 horas que el conjunto de datos (por ejemplo, 05:00 = 17:00:00):

In [34]:
# Comparar los objetos datetime
# llamando a la función para el lunes por la mañana en Springfield (utilizando spr_general en vez de la tabla df)
genre_weekday(spr_general, 'Monday',"07:00:00","11:00:00" )

genre
pop            781
dance          549
electronic     480
rock           474
hiphop         286
ruspop         186
world          181
rusrap         175
alternative    164
unknown        161
classical      157
metal          120
jazz           100
folk            97
soundtrack      95
Name: genre, dtype: int64

In [35]:
# llamando a la función para el lunes por la mañana en Shelbyville (utilizando shel_general en vez de la tabla df)
genre_weekday(shel_general, 'Monday',"07:00:00","11:00:00" )

genre
pop            218
dance          182
rock           162
electronic     147
hiphop          80
ruspop          64
alternative     58
rusrap          55
jazz            44
classical       40
world           36
rap             32
soundtrack      31
rnb             27
metal           27
Name: genre, dtype: int64

In [36]:
# llamando a la función para el viernes por la tarde en Springfield
genre_weekday(spr_general, 'Friday',"17:00:00","23:00:00" )

genre
pop            713
rock           517
dance          495
electronic     482
hiphop         273
world          208
ruspop         170
classical      163
alternative    163
rusrap         142
jazz           111
unknown        110
soundtrack     105
rnb             90
metal           88
Name: genre, dtype: int64

In [37]:
# llamando a la función para el viernes por la tarde en Shelbyville
genre_weekday(shel_general, 'Friday',"17:00:00","23:00:00" )

genre
pop            256
rock           216
electronic     216
dance          210
hiphop          97
alternative     63
jazz            61
classical       60
rusrap          59
world           54
unknown         47
ruspop          47
soundtrack      40
metal           39
rap             36
Name: genre, dtype: int64

**Conclusión**
Es posible rechazar esta hipótesis ya que al analizar los resultados se observa que:
Los lunes por la mañana los usuarios escuchan 12/15 géneros similares en proporción Shelbiville:Springfielg y que los viernes por la noche son 13/15 géneros similares en Shelbiville:Springfielg.
Por lo tanto, los oyentes de ambas ciudades escuchan tanto lunes y viernes géneros musicales similares.

[Volver a Contenidos](#back)

### Hipótesis 3: preferencias de género en Springfield y Shelbyville <a id='genre'></a>

Hipótesis: Shelbyville ama la música rap. A los residentes de Springfield les gusta más el pop.

Agrupa la tabla `spr_general` por género y encuentra el número de canciones reproducidas de cada género con el método `count()`. Después, ordena el resultado en orden descendente y guárdalo en `spr_genres`.

In [38]:
# Escribe una línea de código que:
# 1. agrupe la tabla spr_general por la columna 'genre';
spr_genres = spr_general.groupby('genre')['genre'].count() # 2. cuente los valores 'genre' con count() en la agrupación;
spr_genres = spr_general.sort_values(by= 'genre', ascending=False)


Muestra las 10 primeras filas de `spr_genres`:

In [39]:
# mostrar las 10 primeras filas de spr_genres
display(spr_genres.head(10))

user_id                    track            artist      genre  \
8448   A439123F                  Flip It           unknown        ïîï   
27939  D7FB50DA          Drumming Circle  Professor Trance  worldbeat   
15126  F1AFA4BA             Mae Pa Fidje   Teofilo Chantre      world   
25455  6B799FBC                      Eba           Tamaley      world   
43076  54F93A25                Franc Jeu      Medhy Custos      world   
43050  4C122665  Masha Allah Allah Allah          DJ Nabil      world   
5892   382BE6A7                 Yerazoum  Mihran Tsarukyan      world   
55182  12B1B81D              Nós e o Mar        Tamba Trio      world   
18578  9097998C              Sandy Beach            Broque      world   
5897   7FB00AC8                  '74 '75           The Luc      world   

              city      time        day  
8448   Springfield  09:08:51     Friday  
27939  Springfield  09:30:47     Monday  
15126  Springfield  14:45:29     Friday  
25455  Springfield  21:45:47     Monday  
43076  Springfield  21:41:11     Friday  
43050  Springfield  09:07:44  Wednesday  
5892   Springfield  08:57:02     Monday  
55182  Springfield  13:46:20     Monday  
18578  Springfield  21:12:05     Friday  
5897   Springfield  09:05:33     Friday

Ahora haz lo mismo con los datos de Shelbyville.

Agrupa la tabla `shel_general` por género y encuentra el número de canciones reproducidas de cada género. Después, ordena el resultado en orden descendente y guárdalo en la tabla `shel_genres`:


In [40]:
# Escribe una línea de código que:
# 1. agrupe la tabla shel_general por la columna 'genre';
# 2. cuente los valores 'genre' en el agrupamiento con count();
# 3. ordene el Series resultante en orden descendente y lo guarde en shel_genres.
shel_genres = shel_general.groupby('genre')['genre'].count()
shel_genres = shel_general.sort_values(by= 'genre', ascending=False)

Muestra las 10 primeras filas de `shel_genres`:

In [41]:
# Muestra las 10 primeras filas de shel_genres
display(shel_genres.head(10))

user_id                track                     artist      genre  \
39063  AA1730E8                  Anu          Kailash Kokopelli  worldbeat   
30556  4F773441                Tibet                 Furunkulus      world   
13044  BA0FE104                 Ride            Elji Beatzkilla      world   
38538  38EF7CEB       El Forga Moura                       Haim      world   
19650  58EA4E8A  King Of The Fairies                Tom McHaile      world   
38594  2297CBFB          Minor Swing  International String Trio      world   
1226    47E5088       Massachussetts              Cover Masters      world   
13041  F0F48477          Mustt Mustt                    unknown      world   
5390   1FE0862E      Sakura Shamisen                     Kensei      world   
57717   CAB3F3A                Dovdu                    Angelit      world   

              city      time        day  
39063  Shelbyville  20:16:34  Wednesday  
30556  Shelbyville  21:07:29  Wednesday  
13044  Shelbyville  13:53:55  Wednesday  
38538  Shelbyville  13:43:19     Monday  
19650  Shelbyville  14:47:52  Wednesday  
38594  Shelbyville  14:58:22  Wednesday  
1226   Shelbyville  21:02:08     Friday  
13041  Shelbyville  13:18:22     Friday  
5390   Shelbyville  08:51:08     Friday  
57717  Shelbyville  08:59:59  Wednesday

**Conclusión**
Al analizar los resultados es posible rechazar parcialmente la hipótesis, debido a que los géneros musicales más escuchados tanto en Springfield como en Shelbyville son wordbeat (que fuciona musica tradicional y folklórica con electrónica, rock e influencias del pop) y world (que se compone de músicas no occidentales-tradionales o fusiones con otros estilos no occidentales), por lo que no es posible inferir que los oyentes de esas ciudades les guste específicamente el pop o la música rap.

[Volver a Contenidos](#back)

# Conclusiones <a id='end'></a>

A lo largo de este proyecto se tuvo como objetivo analizar los datos entregados para poder afirmar o rechazar ciertas hipótesis que nos fueron entregadas.

Hipótesis 1: existen diferencias en la forma en que los usuarios y las usuarias de Springfield y Shelbyville consumen música. 

Según los resultados podría afirmarse que los/las usuarias consumen música durante la semana de manera distinta dependiendo a la ciudad.
Shelbyville inicia la semana escuchando menos música en comparación con los días miércoles y viernes, el día que más reproducen música son los miércoles, luego el viernes disminuye un poco. En cambio Springfield tanto los lunes como los viernes hay un alto consumo de música, pero aumenta levemente los viernes; a mitad de la semana (miércoles) el consumo de música disminuye.

Hipótesis 2: el lunes por la mañana y el viernes por la noche, los ciudadanos de Springfield escuchan géneros diferentes a los que disfrutan los usuarios de Shelbyville.

Es posible rechazar esta hipótesis ya que al analizar los resultados se observó que ambas ciudades comparten varios géneros musicales en común, se observó una proporción 12:15 y 13:15.

Hipótesis 3: Shelbyville ama la música rap. A los residentes de Springfield les gusta más el pop.

Es posible rechazar parcialmente la hipótesis, debido a que los géneros musicales más escuchados tanto en Springfield como en Shelbyville son wordbeat y world, por lo que no es posible inferir que los oyentes de esas ciudades les guste específicamente el pop o la música rap. Es necesario más análisis y datos.

### Nota
En proyectos de investigación reales, la prueba de hipótesis estadística es más precisa y cuantitativa. También ten en cuenta que no siempre se pueden sacar conclusiones sobre una ciudad entera a partir de datos de una sola fuente.

Aprenderás más sobre la prueba de hipótesis en el sprint de análisis estadístico de datos.

[Volver a Contenidos](#back)